examples_R10/R10_static_server.c

```diff
@@ -0,0 +1,302 @@
+        "<div class=\"container\">\n"
+        "</div>\n"
+        "</nav>\n"
+        "<main>\n"
+        "<div class=\"container p-3\">\n"
+        "<h2>%s</h2>\n"
+        "<ul class=\"list-unstyled my-3\">\n",
+        decoded_path, /* XXX html-escape this. */
+        path,         /* XXX html-escape this? */
+        trailing_slash, decoded_path /* XXX html-escape this */);
+    while ((ent = readdir(d))) {
+      const char *name = ent->d_name;
+      if (strcmp(name, "."))
+        evbuffer_add_printf(evb, "<li><a href=\"%s\">%s</a>\n", name,
+                            name); /* XXX escape this */
+    }
+    evbuffer_add_printf(evb, "</ul>"
+                             "</div>\n"
+                             "</main>\n"
+                             "<footer class=\"p-3\">\n"
+                             "<div class=\"container\">\n"
+                             "</div>\n"
+                             "</footer>\n"
+                             "</body>"
+                             "</html>\n");
+    closedir(d);
+
+    add_content_length(req, evbuffer_get_length(evb));
+    if (cmd == EVHTTP_REQ_HEAD)
+      evbuffer_drain(evb, evbuffer_get_length(evb));
+    evhttp_add_header(evhttp_request_get_output_headers(req), "Content-Type",
+                      "text/html");
+  } else {
+    /* Otherwise it's a file; add it to the buffer to get
+     * sent via sendfile */
+    if (type == NULL)
+      type = guess_content_type(whole_path);
+    evhttp_add_header(evhttp_request_get_output_headers(req), "Content-Type",
+                      type);
+
+    if (st.st_size != 0) {
+      if ((fd = open(whole_path, O_RDONLY)) == -1) {
+        if (errno == ENOENT) {
+          fprintf(stderr, "File '%s' not found\n", whole_path);
+          evhttp_send_error(req, HTTP_NOTFOUND, NULL);
+        } else {
+          evhttp_send_error(req, HTTP_INTERNAL, NULL);
+        }
+        goto done;
+      }
+      if (cmd != EVHTTP_REQ_HEAD) {
+        if (evbuffer_add_file(evb, fd, 0, st.st_size) != 0) {
+          evhttp_send_error(req, HTTP_INTERNAL, NULL);
+          goto cleanup;
+        }
+      }
+    }
+    add_content_length(req, st.st_size);
+  }
+  evhttp_send_reply(req, HTTP_OK, "OK", evb);
+
+  goto done;
+
+err:
+  evhttp_send_error(req, HTTP_NOTFOUND, NULL);
+cleanup:
+  if (fd >= 0)
+    close(fd);
+
+done:
+  if (decoded)
+    evhttp_uri_free(decoded);
+  if (decoded_path)
+    free(decoded_path);
+  if (evb)
+    evbuffer_free(evb);
+}
+
+int main() {
+  ev_uint16_t http_port = 8080;
+  char *http_addr = "0.0.0.0";
+  struct event_base *base;
+  struct evhttp *http_server;
+
+  base = event_base_new();
+
+  http_server = evhttp_new(base);
+  evhttp_bind_socket(http_server, http_addr, http_port);
+  evhttp_set_gencb(http_server, send_file_to_user, NULL);
+
+  printf("Listening requests on http://%s:%d\n", http_addr, http_port);
+
+  event_base_dispatch(base);
+
+  evhttp_free(http_server);
```

Comment:
It will never enter this code, let's add signal handler to clean gracefully

Reply:
Added for both servers

Reply:
I've fixed one tiny problem

```
R10_static_server.c:319:38: error: ‘SIGINT’ undeclared (first use in this function)
  319 |         sig_int = evsignal_new(base, SIGINT, signal_cb, base);
```